// eslint-config-openlayers/react.js
module.exports = {
  extends: ['prettier', 'prettier/react'],
  plugins: ['prettier', 'react'],
  env: {
    browser: true,
    es6: true
  },
  parserOptions: {
    ecmaVersion: '2017',
    sourceType: 'module',
    ecmaFeatures: {
      experimentalObjectRestSpread: true,
      jsx: true
    }
  },
  rules: {
    'array-callback-return': 'error',
    'block-scoped-var': 'error',
    'curly': 'error',
    'default-case': 'error',
    'dot-notation': ['error', {allowPattern: '^[a-z]+(_[a-z]+)+$'}],
    'eqeqeq': 'error',
    'no-case-declarations': 'error',
    'no-cond-assign': 'error',
    'no-console': 'error',
    'no-control-regex': 'error',
    'no-debugger': 'error',
    'no-delete-var': 'error',
    'no-dupe-args': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty-character-class': 'error',
    'no-empty': 'error',
    'no-eq-null': 'error',
    'no-ex-assign': 'error',
    'no-extra-boolean-cast': 'error',
    'no-fallthrough': 'error',
    'no-func-assign': 'error',
    'no-inner-declarations': ['error', 'functions'],
    'no-invalid-regexp': 'error',
    'no-irregular-whitespace': 'error',
    'no-negated-in-lhs': 'error',
    'no-obj-calls': 'error',
    'no-octal': 'error',
    'no-param-reassign': ['error', {props: true}],
    'no-process-exit': 'off',
    'no-redeclare': 'error',
    'no-regex-spaces': 'error',
    'no-sparse-arrays': 'error',
    'no-undef': 'error',
    'no-underscore-dangle': 'off',
    'no-unexpected-multiline': 'error',
    'no-unreachable': 'error',
    'no-unsafe-finally': 'error',
    'no-unused-vars': ['error', {ignoreRestSiblings: true}],
    'no-use-before-define': ['error', 'nofunc'],
    'prefer-const': 'error',
    'prettier/prettier': ['error', {singleQuote: true, bracketSpacing: false}],
    'react/jsx-boolean-value': 'error',
    'react/jsx-key': 'error',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-undef': 'error',
    'react/jsx-sort-props': 'off',
    'react/jsx-uses-react': 'error',
    'react/jsx-uses-vars': 'error',
    'react/no-array-index-key': 'error',
    'react/no-children-prop': 'error',
    'react/no-did-mount-set-state': 'error',
    'react/no-did-update-set-state': 'error',
    'react/no-direct-mutation-state': 'error',
    'react/no-unescaped-entities': 'error',
    'react/no-unknown-property': 'error',
    'react/no-unused-prop-types': 'error',
    'react/prop-types': 'error',
    'react/react-in-jsx-scope': 'error',
    'react/sort-prop-types': 'off',
    'strict': 'off',
    'use-isnan': 'error',
    'valid-typeof': 'error'
  }
};
